File: mobile-sdks/ios/NigeriaGeoSDK/Sources/NigeriaGeoSDK/NigeriaGeoConfig.swift
import Foundation

/// Configuration for the Nigeria Geo SDK
public struct NigeriaGeoConfig {
    
    // MARK: - Properties
    
    /// Base URL for the API
    public let baseURL: String
    
    /// Optional API key for authentication
    public let apiKey: String?
    
    /// Whether to enable local caching
    public let enableCaching: Bool
    
    /// Cache expiration duration
    public let cacheExpiration: CacheExpiration
    
    /// Maximum cache size
    public let maxCacheSize: CacheSize
    
    /// Request timeout in seconds
    public let requestTimeout: TimeInterval
    
    /// Whether to enable performance monitoring
    public let enablePerformanceMonitoring: Bool
    
    /// Custom HTTP headers
    public let customHeaders: [String: String]
    
    // MARK: - Initialization
    
    /// Initialize SDK configuration
    /// - Parameters:
    ///   - baseURL: API base URL
    ///   - apiKey: Optional API key
    ///   - enableCaching: Enable local caching (default: true)
    ///   - cacheExpiration: Cache expiration duration (default: 24 hours)
    ///   - maxCacheSize: Maximum cache size (default: 50MB)
    ///   - requestTimeout: Request timeout (default: 30 seconds)
    ///   - enablePerformanceMonitoring: Enable performance monitoring (default: false)
    ///   - customHeaders: Custom HTTP headers (default: empty)
    public init(
        baseURL: String,
        apiKey: String? = nil,
        enableCaching: Bool = true,
        cacheExpiration: CacheExpiration = .hours(24),
        maxCacheSize: CacheSize = .megabytes(50),
        requestTimeout: TimeInterval = 30.0,
        enablePerformanceMonitoring: Bool = false,
        customHeaders: [String: String] = [:]
    ) {
        self.baseURL = baseURL
        self.apiKey = apiKey
        self.enableCaching = enableCaching
        self.cacheExpiration = cacheExpiration
        self.maxCacheSize = maxCacheSize
        self.requestTimeout = requestTimeout
        self.enablePerformanceMonitoring = enablePerformanceMonitoring
        self.customHeaders = customHeaders
    }
}

// MARK: - Cache Configuration

/// Cache expiration options
public enum CacheExpiration {
    case minutes(Int)
    case hours(Int)
    case days(Int)
    case never
    
    var timeInterval: TimeInterval {
        switch self {
        case .minutes(let minutes):
            return TimeInterval(minutes * 60)
        case .hours(let hours):
            return TimeInterval(hours * 3600)
        case .days(let days):
            return TimeInterval(days * 24 * 3600)
        case .never:
            return TimeInterval.greatestFiniteMagnitude
        }
    }
}

/// Cache size options
public enum CacheSize {
    case kilobytes(Int)
    case megabytes(Int)
    case gigabytes(Int)
    
    var bytes: Int {
        switch self {
        case .kilobytes(let kb):
            return kb * 1024
        case .megabytes(let mb):
            return mb * 1024 * 1024
        case .gigabytes(let gb):
            return gb * 1024 * 1024 * 1024
        }
    }
}

// MARK: - Default Configurations

public extension NigeriaGeoConfig {
    
    /// Default configuration for development
    static func development(baseURL: String, apiKey: String? = nil) -> NigeriaGeoConfig {
        return NigeriaGeoConfig(
            baseURL: baseURL,
            apiKey: apiKey,
            enableCaching: true,
            cacheExpiration: .hours(1), // Shorter cache for development
            maxCacheSize: .megabytes(10),
            requestTimeout: 15.0,
            enablePerformanceMonitoring: true
        )
    }
    
    /// Default configuration for production
    static func production(baseURL: String, apiKey: String? = nil) -> NigeriaGeoConfig {
        return NigeriaGeoConfig(
            baseURL: baseURL,
            apiKey: apiKey,
            enableCaching: true,
            cacheExpiration: .hours(24),
            maxCacheSize: .megabytes(100),
            requestTimeout: 30.0,
            enablePerformanceMonitoring: false
        )
    }
    
    /// Configuration with minimal caching for testing
    static func testing(baseURL: String, apiKey: String? = nil) -> NigeriaGeoConfig {
        return NigeriaGeoConfig(
            baseURL: baseURL,
            apiKey: apiKey,
            enableCaching: false,
            cacheExpiration: .minutes(1),
            maxCacheSize: .megabytes(1),
            requestTimeout: 10.0,
            enablePerformanceMonitoring: true
        )
    }
}

File: mobile-sdks/flutter/lib/src/client/nigeria_geo_config.dart
class NigeriaGeoConfig {
  /// Base URL for the Nigeria Geo API
  final String baseUrl;

  /// API key for authentication (optional)
  final String? apiKey;

  /// Whether to enable caching
  final bool enableCaching;

  /// Cache expiration duration
  final Duration cacheExpiration;

  /// Maximum cache size in bytes
  final int maxCacheSize;

  /// Request timeout duration
  final Duration timeout;

  /// Whether to enable detailed logging
  final bool enableLogging;

  /// Custom user agent for HTTP requests
  final String? userAgent;

  /// Whether to enable retry on network failures
  final bool enableRetry;

  /// Maximum number of retry attempts
  final int maxRetries;

  /// Base delay between retry attempts
  final Duration retryDelay;

  const NigeriaGeoConfig({
    required this.baseUrl,
    this.apiKey,
    this.enableCaching = true,
    this.cacheExpiration =
        const Duration(hours: 6), // Shorter cache for fast API
    this.maxCacheSize = 25 * 1024 * 1024, // 25MB - reduced cache size
    this.timeout = const Duration(seconds: 5), // Optimized timeout
    this.enableLogging = false,
    this.userAgent,
    this.enableRetry = true,
    this.maxRetries = 2, // Fewer retries for fast API
    this.retryDelay = const Duration(milliseconds: 500), // Faster retry
  });

  /// Creates a config for development environment
  factory NigeriaGeoConfig.development({
    required String baseUrl,
    String? apiKey,
  }) {
    return NigeriaGeoConfig(
      baseUrl: baseUrl,
      apiKey: apiKey,
      enableCaching: true,
      cacheExpiration: const Duration(minutes: 30),
      enableLogging: true,
      timeout: const Duration(seconds: 10),
      maxRetries: 1,
    );
  }

  /// Creates a config for production environment
  factory NigeriaGeoConfig.production({
    required String baseUrl,
    required String apiKey,
  }) {
    return NigeriaGeoConfig(
      baseUrl: baseUrl,
      apiKey: apiKey,
      enableCaching: true,
      cacheExpiration: const Duration(hours: 24),
      enableLogging: false,
      timeout: const Duration(seconds: 30),
      maxRetries: 3,
      retryDelay: const Duration(seconds: 2),
    );
  }

  /// Creates a config with caching disabled
  factory NigeriaGeoConfig.noCaching({
    required String baseUrl,
    String? apiKey,
  }) {
    return NigeriaGeoConfig(
      baseUrl: baseUrl,
      apiKey: apiKey,
      enableCaching: false,
      enableLogging: false,
    );
  }

  /// Copy config with modified values
  NigeriaGeoConfig copyWith({
    String? baseUrl,
    String? apiKey,
    bool? enableCaching,
    Duration? cacheExpiration,
    int? maxCacheSize,
    Duration? timeout,
    bool? enableLogging,
    String? userAgent,
    bool? enableRetry,
    int? maxRetries,
    Duration? retryDelay,
  }) {
    return NigeriaGeoConfig(
      baseUrl: baseUrl ?? this.baseUrl,
      apiKey: apiKey ?? this.apiKey,
      enableCaching: enableCaching ?? this.enableCaching,
      cacheExpiration: cacheExpiration ?? this.cacheExpiration,
      maxCacheSize: maxCacheSize ?? this.maxCacheSize,
      timeout: timeout ?? this.timeout,
      enableLogging: enableLogging ?? this.enableLogging,
      userAgent: userAgent ?? this.userAgent,
      enableRetry: enableRetry ?? this.enableRetry,
      maxRetries: maxRetries ?? this.maxRetries,
      retryDelay: retryDelay ?? this.retryDelay,
    );
  }

  @override
  String toString() {
    return 'NigeriaGeoConfig('
        'baseUrl: $baseUrl, '
        'hasApiKey: ${apiKey != null}, '
        'enableCaching: $enableCaching, '
        'cacheExpiration: $cacheExpiration, '
        'timeout: $timeout'
        ')';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is NigeriaGeoConfig &&
        other.baseUrl == baseUrl &&
        other.apiKey == apiKey &&
        other.enableCaching == enableCaching &&
        other.cacheExpiration == cacheExpiration &&
        other.maxCacheSize == maxCacheSize &&
        other.timeout == timeout &&
        other.enableLogging == enableLogging &&
        other.userAgent == userAgent &&
        other.enableRetry == enableRetry &&
        other.maxRetries == maxRetries &&
        other.retryDelay == retryDelay;
  }

  @override
  int get hashCode {
    return Object.hash(
      baseUrl,
      apiKey,
      enableCaching,
      cacheExpiration,
      maxCacheSize,
      timeout,
      enableLogging,
      userAgent,
      enableRetry,
      maxRetries,
      retryDelay,
    );
  }
}
